templates/c/src/tic80.h

```diff
@@ -0,0 +1,208 @@
+const uint32_t TILES_SIZE = 8192;
+const uint32_t SPRITES_SIZE = 8192;
+const uint32_t MAP_SIZE = 32640;
+const uint32_t GAMEPADS_SIZE = 4;
+const uint32_t MOUSE_SIZE = 4;
+const uint32_t KEYBOARD_SIZE = 4;
+const uint32_t SFX_STATE_SIZE = 16;
+const uint32_t SOUND_REGISTERS_SIZE = 72;
+const uint32_t WAVEFORMS_SIZE = 256;
+const uint32_t SFX_SIZE = 4224;
+const uint32_t MUSIC_PATTERNS_SIZE = 11520;
+const uint32_t MUSIC_TRACKS_SIZE = 408;
+const uint32_t SOUND_STATE_SIZE = 4;
+const uint32_t STEREO_VOLUME_SIZE = 4;
+const uint32_t PERSISTENT_MEMORY_SIZE = 1024;
+const uint32_t SPRITE_FLAGS_SIZE = 512;
+const uint32_t SYSTEM_FONT_SIZE = 2048;
+const uint32_t WASM_FREE_RAM_SIZE = 163840; // 160kb
+
+// Pointers.
+VRAM* FRAMEBUFFER = (VRAM*)0;
+uint8_t* TILES = (uint8_t*)0x04000;
+uint8_t* SPRITES = (uint8_t*)0x06000;
+uint8_t* MAP = (uint8_t*)0x08000;
+uint8_t* GAMEPADS = (uint8_t*)0x0FF80;
+uint8_t* MOUSE = (uint8_t*)0x0FF84;
+uint8_t* KEYBOARD = (uint8_t*)0x0FF88;
+uint8_t* SFX_STATE = (uint8_t*)0x0FF8C;
+uint8_t* SOUND_REGISTERS = (uint8_t*)0x0FF9C;
+uint8_t* WAVEFORMS = (uint8_t*)0x0FFE4;
+uint8_t* SFX = (uint8_t*)0x100E4;
+uint8_t* MUSIC_PATTERNS = (uint8_t*)0x11164;
+uint8_t* MUSIC_TRACKS = (uint8_t*)0x13E64;
+uint8_t* SOUND_STATE = (uint8_t*)0x13FFC;
+uint8_t* STEREO_VOLUME = (uint8_t*)0x14000;
+uint8_t* PERSISTENT_MEMORY = (uint8_t*)0x14004;
+uint8_t* SPRITE_FLAGS = (uint8_t*)0x14404;
+uint8_t* SYSTEM_FONT = (uint8_t*)0x14604;
+uint8_t* WASM_FREE_RAM = (uint8_t*)0x18000; // 160kb
+
+// Functions.
+WASM_IMPORT("btn")
+int32_t btn(int32_t id);
+
+WASM_IMPORT("btnp")
+bool btnp(int32_t id, int32_t hold, int32_t period);
+
+WASM_IMPORT("circ")
+void circ(int32_t x, int32_t y, int32_t radius, int32_t color);
+
+WASM_IMPORT("circb")
+void circb(int32_t x, int32_t y, int32_t radius, int32_t color);
+
+WASM_IMPORT("clip")
+void clip(int32_t x, int32_t y, int32_t w, int32_t h);
+
+WASM_IMPORT("cls")
+void cls(int32_t color);
+
+WASM_IMPORT("exit")
+void exit();
+
+WASM_IMPORT("elli")
+void elli(int32_t x, int32_t y, int32_t a, int32_t b, int32_t color);
+
+WASM_IMPORT("ellib")
+void ellib(int32_t x, int32_t y, int32_t a, int32_t b, int32_t color);
+
+WASM_IMPORT("fget")
+bool fget(int32_t id, uint8_t flag);
+
+WASM_IMPORT("font")
+int32_t font(const char* text, int32_t x, int32_t y, uint32_t* transcolors, int32_t colorcount, int32_t width, int32_t height, bool fixed, int32_t scale, bool alt);
+
+WASM_IMPORT("fset")
+bool fset(int32_t id, uint8_t flag, bool value);
+
+WASM_IMPORT("key")
+bool key(int32_t keycode);
+
+WASM_IMPORT("keyp")
+bool keyp(int32_t keycode, int32_t hold, int32_t period);
+
+WASM_IMPORT("line")
+void line(float x0, float y0, float x1, float y1, int8_t color);
+
+WASM_IMPORT("map")
+void map(int32_t x, int32_t y, int32_t w, int32_t h, int32_t sx, int32_t sy, uint32_t* transcolors, int32_t colorcount, int32_t scale, int32_t remap);
+
+WASM_IMPORT("memcpy")
+void memcpy_tic(uint32_t copyto, uint32_t copyfrom, uint32_t length);
```

Comment:
You make a good case...

Reply:
As pointed out elsewhere C's own native `memcpy` would likely be sufficient - there isn't supposed to be any magic.... is tapping into that a possibility?

Reply:
I didn't know that was allowed for this, I assumed TIC's memcpy and memset were special.

Reply:
They shouldn't be.  The WASM VM has access to all the RAM - it should be able to move it about at at will .  Really the `memcpy` and `memset` make a whole lot less sense when you're in C as you have directly access to RAM.  Remember most people are using scripting runtime so the only way to get to the devices  96kb of MMIO is via these special functions, where-as in the WASM runtime all the memory is just freely available.

Reply:
Unless the API is different though, that's a sticky point... we'd prefer the same TIC API, but if we used native memory copy I thhink that'd be fine.

Reply:
Lets just add a mention to the README that for simplicity and to avoid conflicts one should use the standard C memory functions rather than the TIC api ones...

Reply:
I have updated the README. I think the wording is ok, but I'm not sure.